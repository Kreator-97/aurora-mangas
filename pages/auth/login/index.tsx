import { FormEvent } from 'react'
import { GetServerSideProps } from 'next'
import { useRouter } from 'next/router'
import Link from 'next/link'
import { FcGoogle } from 'react-icons/fc'
import { HiArrowSmLeft } from 'react-icons/hi'
import { getSession, signIn } from 'next-auth/react'

import { useForm } from '../../../hooks'

const LoginPage = () => {
  const router = useRouter()
  const { email, password, onInputChange } = useForm({
    email: '',
    password: '',
  })

  const onGoogleSignIn = () => {
    try {
      signIn('google')
    } catch(error) {
      console.log(error)
    }
  }

  const onSubmit = (e:FormEvent) => {
    e.preventDefault()
    const required = [email, password]
    if( required.some( value => value.trim() === '' || !value) ) return

    signIn('credentials', { email, password })
  }

  return (
    <div className="mx-auto max-w-lg min-h-screen flex items-center md:w-1/2">

      <form
        className="bg-light bg-opacity-90 w-full p-4 h-screen sm:h-fit" onSubmit={ onSubmit }
        data-testid="form-login"
      >
        <div className='flex justify-between mb-2'>
          <HiArrowSmLeft fill='#444' size={28} cursor="pointer" onClick={ () => router.back() }/>
          <h1 className="text-2xl text-center text-dark">Iniciar sesión</h1>
          <div className='w-7'></div>
        </div>

        <div className='mb-2'>
          <label htmlFor="email" className="text-dark">
          Ingresa tu correo electrónico:
            <input
              className="w-full bg-light bg-opacity-80 px-4 py-2 text-dark border border-stroke border-solid focus:outline-none mb-2"
              placeholder="Correo electrónico"
              type="email"
              name="email"
              value={ email }
              onChange={ onInputChange }
            />
          </label>
        </div>

        <div className='mb-2'>
          <label htmlFor="password" className="text-dark">
          Ingresa tu contraseña:
            <input
              className="w-full bg-light bg-opacity-80 px-4 py-2 text-dark border border-stroke border-solid focus:outline-none mb-2"
              placeholder="Contraseña"
              type="password"
              name="password"
              value={ password }
              onChange={ onInputChange }
              data-testid="password"
            />
          </label>
        </div>
        <button
          className='btn bg-accent hover:bg-accentDark w-full mb-2'
          type="submit"
        >Iniciar sesión</button>
        
        <h2 className="text-center text-lg text-dark mb-2">Inicia sesión con tu cuenta Google:</h2>
        <div className='flex justify-center mb-4'>
          <FcGoogle
            size={48}
            cursor="pointer"
            onClick={() => onGoogleSignIn() }
            data-testid="google-icon"
          />
        </div>

        <Link href={'/auth/register'} passHref>
          <a className='block text-center btn bg-accent hover:bg-accentDark w-full ghost'>
          ¿No tienes cuenta?, Registrate fácilmente
          </a>
        </Link>
      </form>
    </div>
  )
}

export default LoginPage

export const getServerSideProps: GetServerSideProps = async (ctx) => {
  const session = await getSession(ctx)

  if( session ) {
    return {
      redirect: {
        destination: '/',
        permanent: false,
      }
    }
  }

  return {
    props: {}
  }
}
