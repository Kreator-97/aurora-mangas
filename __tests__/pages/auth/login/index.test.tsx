import { fireEvent, render, screen } from '@testing-library/react'
import { default as LoginPage } from '../../../../pages/auth/login/index'

import { signIn } from 'next-auth/react'

jest.mock('next-auth/react')

describe('tests on LoginPage', () => {
  test('should to match with snapshot', () => {
    const { container } = render(<LoginPage />)
    expect(container).toMatchSnapshot()
    expect(screen.getByRole('heading', { name: 'Iniciar sesión' })).toBeInTheDocument
    expect(screen.getByText('Ingresa tu correo electrónico:')).toBeInTheDocument
    expect(screen.getByText('Ingresa tu contraseña:')).toBeInTheDocument
  })

  test('should to call signIn()', () => {
    (signIn as jest.Mock).mockReturnValue(jest.fn())

    render(<LoginPage />)
    const email = 'correo@correo.com', password = '12345678'
    const form = screen.getByTestId('form-login')
    const emailInput = screen.getByRole('textbox', {name: ''})
    const passwordInput = screen.getByTestId('password')
    
    fireEvent.change(emailInput, {target: { name: 'email', value: email}})
    fireEvent.change(passwordInput, {target: { name: 'password', value: password }})
    fireEvent.submit(form)
    
    expect(signIn).toHaveBeenCalledWith('credentials', {email, password})
  })

  test('should to call signIn with google', () => {
    (signIn as jest.Mock).mockReturnValue(jest.fn())

    render(<LoginPage />)

    const googleIcon = screen.getByTestId('google-icon')

    fireEvent.click(googleIcon)

    expect(signIn).toHaveBeenCalledWith('google')
  })
})
