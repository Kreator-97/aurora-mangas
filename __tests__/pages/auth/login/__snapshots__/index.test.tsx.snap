// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`tests on LoginPage should to match with snapshot 1`] = `
<div>
  <div
    class="mx-auto max-w-lg min-h-screen flex items-center md:w-1/2"
  >
    <form
      class="bg-light bg-opacity-90 w-full p-4 h-screen sm:h-fit"
      data-testid="form-login"
    >
      <div
        class="flex justify-between mb-2"
      >
        <svg
          cursor="pointer"
          fill="#444"
          height="28"
          stroke="currentColor"
          stroke-width="0"
          viewBox="0 0 20 20"
          width="28"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            clip-rule="evenodd"
            d="M9.707 14.707a1 1 0 01-1.414 0l-4-4a1 1 0 010-1.414l4-4a1 1 0 011.414 1.414L7.414 9H15a1 1 0 110 2H7.414l2.293 2.293a1 1 0 010 1.414z"
            fill-rule="evenodd"
          />
        </svg>
        <h1
          class="text-2xl text-center text-dark"
        >
          Iniciar sesión
        </h1>
        <div
          class="w-7"
        />
      </div>
      <div
        class="mb-2"
      >
        <label
          class="text-dark"
          for="email"
        >
          Ingresa tu correo electrónico:
          <input
            class="w-full bg-light bg-opacity-80 px-4 py-2 text-dark border border-stroke border-solid focus:outline-none mb-2"
            name="email"
            placeholder="Correo electrónico"
            type="email"
            value=""
          />
        </label>
      </div>
      <div
        class="mb-2"
      >
        <label
          class="text-dark"
          for="password"
        >
          Ingresa tu contraseña:
          <input
            class="w-full bg-light bg-opacity-80 px-4 py-2 text-dark border border-stroke border-solid focus:outline-none mb-2"
            data-testid="password"
            name="password"
            placeholder="Contraseña"
            type="password"
            value=""
          />
        </label>
      </div>
      <button
        class="btn bg-accent hover:bg-accentDark w-full mb-2"
        type="submit"
      >
        Iniciar sesión
      </button>
      <h2
        class="text-center text-lg text-dark mb-2"
      >
        Inicia sesión con tu cuenta Google:
      </h2>
      <div
        class="flex justify-center mb-4"
      >
        <svg
          cursor="pointer"
          data-testid="google-icon"
          enable-background="new 0 0 48 48"
          fill="currentColor"
          height="48"
          stroke="currentColor"
          stroke-width="0"
          version="1.1"
          viewBox="0 0 48 48"
          width="48"
          x="0px"
          xmlns="http://www.w3.org/2000/svg"
          y="0px"
        >
          <path
            d="M43.611,20.083H42V20H24v8h11.303c-1.649,4.657-6.08,8-11.303,8c-6.627,0-12-5.373-12-12
	c0-6.627,5.373-12,12-12c3.059,0,5.842,1.154,7.961,3.039l5.657-5.657C34.046,6.053,29.268,4,24,4C12.955,4,4,12.955,4,24
	c0,11.045,8.955,20,20,20c11.045,0,20-8.955,20-20C44,22.659,43.862,21.35,43.611,20.083z"
            fill="#FFC107"
          />
          <path
            d="M6.306,14.691l6.571,4.819C14.655,15.108,18.961,12,24,12c3.059,0,5.842,1.154,7.961,3.039l5.657-5.657
	C34.046,6.053,29.268,4,24,4C16.318,4,9.656,8.337,6.306,14.691z"
            fill="#FF3D00"
          />
          <path
            d="M24,44c5.166,0,9.86-1.977,13.409-5.192l-6.19-5.238C29.211,35.091,26.715,36,24,36
	c-5.202,0-9.619-3.317-11.283-7.946l-6.522,5.025C9.505,39.556,16.227,44,24,44z"
            fill="#4CAF50"
          />
          <path
            d="M43.611,20.083H42V20H24v8h11.303c-0.792,2.237-2.231,4.166-4.087,5.571
	c0.001-0.001,0.002-0.001,0.003-0.002l6.19,5.238C36.971,39.205,44,34,44,24C44,22.659,43.862,21.35,43.611,20.083z"
            fill="#1976D2"
          />
        </svg>
      </div>
      <a
        class="block text-center btn bg-accent hover:bg-accentDark w-full ghost"
        href="/auth/register"
      >
        ¿No tienes cuenta?, Registrate fácilmente
      </a>
    </form>
  </div>
</div>
`;
